docs/workflows/best-practices-architecture.md

```diff
@@ -0,0 +1,76 @@
+
+<img src="../images/methodology/components-example-1.png" width="500" />
+
+Defining components in this way is perfectly fine. Using a reference modularization pattern also has a major benefit: Anyone who knows the pattern will immediately understand the modularization of your system. People will understand where new functionality should be implemented, since every layer has clearly defined boundaries. 
+
+Defining components based on technical layers does have limitations when the system continues to evolve and grow. One of the limitations is that the number of layers is fixed. If the amount of functionality keeps growing without revisiting the architecture, you can easily end up with a situation where you have 100+ controllers, 100+ services, and 100+ repositories. 
+
+That brings us to a more serious limitation: The component structure does not introduce any restrictions on dependencies beyond the layer restrictions. Sure, controllers should only call services, and services should only call repositories. But any controller can call any service, and that is going to be an issue once you have 100+ controllers and 100+ services. At some point you will also need to introduce functional restrictions in terms of which controllers can access which services. Otherwise, you’re going to end up with uncontrolled dependencies between the layers. 
+
+<img src="../images/methodology/components-example-2.png" width="500" />
+
+Another, though less severe, issue is that nearly every change is going to touch multiple components. You will get the occasional bugfix that only affects a service, but most changes will touch the controller, service, and repository within a certain functional area. 
+
+This can be addressed by introducing functional components, and then retaining the controller-service-repository structure within those functional components. 
+
+<img src="../images/methodology/components-example-3.png" width="500" />
+
+This effectively adds a level of abstraction. You still get the clearly defined roles of the controllers, services, and repositories. But this adds a functional structure on top of that technical structure. This allows you to track and manage dependencies between functional components. Also, since most changes relate to functional area, this means that changes have a more clearly defined scope. 
+
+When it comes to defining components, consistency can be a double-edged sword. Defining functional components has benefits, but it’s almost impossible to define such a structure for your entire system. You’re almost inevitably going to end up with a `common` or `shared` or `utility` component that contains cross-cutting concerns used by all functional components. The outcome is a mixed structure, where most of your components are defined based on functional boundaries, but you also have a small number of technical components. This is fine, having 100% consistency across all components is not a goal in itself.
```

Comment:
I think the point made is very valid, but framed a bit negatively: I would just point out that you cannot _only_ rely on a linear/hierarchical decomposition based on functionality, and that there will also be additional components based on either functional (e.g. taxes, stats collection, resource usage, ..) and non-functional (e.g. authorization, logging, transaction and data management) repetitive or crosscutting concerns.

Reply:
I was actually trying to make the point that 100% consistency is impossible, and you shouldn't make your component structure *less* clear due to consistency. I'll add your examples to make that more concrete.